transformer_engine/common/dropout/dropout.cu

```diff
@@ -0,0 +1,284 @@
+/*************************************************************************
+ * Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#include <curand.h>
+#include <curand_kernel.h>
+#include <curand_philox4x32_x.h>
+
+#include "../common.h"
+#include "transformer_engine/dropout.h"
+
+namespace transformer_engine {
+
+using ull = unsigned long long;
+struct ull2 {
+  unsigned long long x;
+  unsigned long long y;
+};
+
+inline __device__ uint16_t packed_le_8bit(unsigned A, unsigned B) {
+  unsigned T = A & 0x7F7F7F7F;
+  unsigned D = B | 0x80808080;
+  D = -T + D;
+  T = (A ^ B) | 0x7F7F7F7F;
+  D = ~(T ^ D);
+  asm("lop3.b32 %0, %1, %2, %3, 0x4d;\n\t" : "=r"(D) : "r"(A), "r"(B), "r"(D));
+  D = D & 0x80808080;
+  T = (D & 0x0000FFFF) | (D >> 17);
+  return T;
+}
+template <typename T>
+__global__ void dropout_kernel_fwd(T *input, T *output, uint16_t *mask, int64_t *rng_state,
+                                   int n_rng_blocks, unsigned p_dropout_in_uint32_t,
+                                   float inv_prob) {
+  int tid = (blockIdx.y * gridDim.x + blockIdx.x) * blockDim.x + threadIdx.x;
+  int num_threads = gridDim.y * gridDim.x * blockDim.x;
+  uint16_t result = 0;
+  size_t ofs = tid;
+  curandStatePhilox4_32_10_t state;
+  curand_init(rng_state[0], ofs, rng_state[1], &state);
+  ull2 input8;
+  ull2 *gmem_input_ptr = reinterpret_cast<ull2 *>(input);
+  ull2 *gmem_output_ptr = reinterpret_cast<ull2 *>(output);
+
+  while (tid < n_rng_blocks) {
+    uint4 random_uint4 = curand4(&state);
+    result = packed_le_8bit(random_uint4.x, p_dropout_in_uint32_t);
+    result |= packed_le_8bit(random_uint4.y, p_dropout_in_uint32_t) >> 2;
+    result |= packed_le_8bit(random_uint4.z, p_dropout_in_uint32_t) >> 4;
+    result |= packed_le_8bit(random_uint4.w, p_dropout_in_uint32_t) >> 6;
```

Comment:
@vasunvidia could you give some explanation of what those lines do? Especially the shifts by 2/4/6 are strange to me - I would have expected shifts by 4/8/12 (since you take 4 random values in each line?).

Also an idea of the optimization for Blackwell - you should be able to use cvt.rs (stochastic rounding instruction) to do mostly the same thing as what this `packed_le_8bit` instruction is doing, I think.

Reply:
Each call to packed_le_8bit compares 4 random byte values with p_dropout_in_uint32_t and produces a 16-bit output which has the 4-mask bits in 15,14,7,6 th bit positions. So I shift ith call's output by 2^i to get full 16-bit mask.

Reply:
I've added comments so future maintainers can actually understand this. I've also tweaked the bit shift logic so the bit pattern is less weird.